### PR TITLE
planner: get plan from the general plan cache before optimization

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #36598

Problem Summary: planner: get plan from the general plan cache before optimization

### What is changed and how it works?

planner: get plan from the general plan cache before optimization

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cache_param.go [hunk 1]
```diff
@@ -127,3 +130,23 @@ func RestoreASTWithParams(_ sessionctx.Context, stmt ast.StmtNode, params []*dri
...
+			Value:   p.Datum,
+			RetType: tp,
+		})
+	}
+	return exprs
+}
+
+// Available4GeneralPlanCache checks where this stmt is available for general plan cache.
+func Available4GeneralPlanCache(_ sessionctx.Context, stmt ast.StmtNode, is infoschema.InfoSchema) bool {
+	// TODO: implement this function
```
Comment (qw4990) [inline planner/core/plan_cache_param.go:150]: It'll be implemented in the next PR by @Reminiscent

File: planner/core/plan_cache_test.go [hunk 2]
```diff
@@ -107,3 +109,34 @@ func TestInitLRUWithSystemVar(t *testing.T) {
...
+	}
+
+	queries := []string{
+		"select * from t where a<10",
+		"select * from t where a<13 and b<15",
+		"select * from t where b=13",
+		"select * from t where c<8",
+		"select * from t where d>8",
+		"select * from t where c=8 and d>10",
+		"select * from t where a<12 and b<13 and c<12 and d>2",
```
Comment (qw4990) [inline planner/core/plan_cache_test.go:129]: I'll add more test cases in the following several PRs.

**Review by Yisaer — APPROVED** [review]

File: planner/core/plan_cache_param.go [hunk 3]
```diff
@@ -127,3 +130,23 @@ func RestoreASTWithParams(_ sessionctx.Context, stmt ast.StmtNode, params []*dri
...
+	return true
```
Comment (Reminiscent) [inline planner/core/plan_cache_param.go:151]: Maybe it's better to return false here.

File: planner/optimize.go [hunk 4]
```diff
@@ -133,20 +162,33 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
...
 	}
 
+	// try to get Plan from the General Plan Cache
+	if sctx.GetSessionVars().EnableGeneralPlanCache &&
+		isStmtNode &&
+		!useBinding { // TODO: support binding
+		cachedPlan, names, ok, err := getPlanFromGeneralPlanCache(ctx, sctx, stmtNode, is)
+		if err != nil {
+			return nil, nil, err
+		}
```
Comment (Reminiscent) [inline planner/optimize.go:186]: Should we fall back to the normal path?

**Review by Reminiscent — APPROVED** [review]

File: planner/optimize.go (hunk 4 above)
Comment (qw4990) [reply planner/optimize.go:186]: Actually, I have thought about it, but on second thought, I think throwing the error directly here can help us find more issues when testing.
So I suggest that we let it throw the error directly for now, and then change it to fall-back later after we finish some tests.

**Comment by qw4990** [comment]
Wait for https://github.com/pingcap/tidb/pull/37636

File: session/session.go [hunk 5]
```diff
@@ -147,10 +147,6 @@ type Session interface {
 	ExecuteStmt(context.Context, ast.StmtNode) (sqlexec.RecordSet, error)
 	// Parse is deprecated, use ParseWithParams() instead.
 	Parse(ctx context.Context, sql string) ([]ast.StmtNode, error)
-	// Parameterize tries to convert the general statement to an execute statement and then uses general plan cache to handle it.
-	// e.g. "select * from t where a>23" --> "execute 'select * from t where a>?' using 23".
-	// By using the general plan cache, it can skip the parse and optimization stage so to gain some performance benefits.
-	Parameterize(ctx context.Context, originSQL string) (*ast.ExecuteStmt, bool)
```
Comment (qw4990) [inline session/session.go:153]: It's useless now.

**Review by Reminiscent — APPROVED** [review]

**Pull request merged**